Do not promote MediaStream HTMLMediaElement to be NowPlaying without the web page using the MediaSession API

youennf opened the pull request:
#### 40a84b10af90bfa6acfd1a8506517a304c699381
<pre>
Do not promote MediaStream HTMLMediaElement to be NowPlaying without the web page using the MediaSession API
<a href="https://bugs.webkit.org/show_bug.cgi?id=270522">https://bugs.webkit.org/show_bug.cgi?id=270522</a>
<a href="https://rdar.apple.com/121562032">rdar://121562032</a>

Reviewed by Eric Carlson.

When a web page is becoming NowPlaying, it has to react to Play/Pause commands from the user or from user actions like removing AirPods.
Pausing/Resuming media from regular video streaming is a good default, but pausing/resuming live video streaming (no buffering) is not a great default,
like can be seen in video conferencing or game streaming web sites.

We further restrict MediaStream HTMLMediaElement to becoming NowPlaying to the case where the website is using MediaSession API.
Either it should set metadata or provide action handlers.

* LayoutTests/fast/mediastream/now-playing-and-mediastream-expected.txt: Added.
* LayoutTests/fast/mediastream/now-playing-and-mediastream.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::hasNowPlayingInfo const):
(WebCore::MediaElementSession::playbackStateChanged):
(WebCore::MediaElementSession::actionHandlersChanged):

Canonical link: <a href="https://commits.webkit.org/276028@main">https://commits.webkit.org/276028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92a524958608a9c57c5ef4276d04ec50dc4f4348

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22071 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45447 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45670 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39170 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45351 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25799 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35603 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43618 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19102 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37072 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16590 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16693 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38145 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1101 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39223 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38460 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47264 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17923 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14745 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42415 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19497 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37393 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41072 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9696 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19676 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19127 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/html/MediaElementSession.cpp:
```diff
@@ -1241,14 +1241,16 @@ void MediaElementSession::didReceiveRemoteControlCommand(RemoteControlCommandTyp
 
 bool MediaElementSession::hasNowPlayingInfo() const
 {
```
I think this would greatly reduce the current user interface when something is playing. Right now it shows in the Lock Screen (on iOS) or in the control centre on Mac. Following this it won't show at all.

Isn't there a way to prevent automatic play/pause (such as when removing the AirPods) but would still let you interact with the video controls if you do want to pause or resume ? or even get some details about what is playing?

This seems like a really unfortunate side-effect that just to prevent removing the AirPods to pause the video, to prevent all other interactions while at it

youennf replied:
Without more OS support, it is difficult to make it work properly by default without web page help.

youennf committed:
Do not promote MediaStream HTMLMediaElement to be NowPlaying without the web page using the MediaSession API
https://bugs.webkit.org/show_bug.cgi?id=270522
rdar://121562032

Reviewed by Eric Carlson.

When a web page is becoming NowPlaying, it has to react to Play/Pause commands from the user or from user actions like removing AirPods.
Pausing/Resuming media from regular video streaming is a good default, but pausing/resuming live video streaming (no buffering) is not a great default,
like can be seen in video conferencing or game streaming web sites.

We further restrict MediaStream HTMLMediaElement to becoming NowPlaying to the case where the website is using MediaSession API.
Either it should set metadata or provide action handlers.

* LayoutTests/fast/mediastream/now-playing-and-mediastream-expected.txt: Added.
* LayoutTests/fast/mediastream/now-playing-and-mediastream.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::hasNowPlayingInfo const):
(WebCore::MediaElementSession::playbackStateChanged):
(WebCore::MediaElementSession::actionHandlersChanged):

Canonical link: https://commits.webkit.org/276028@main
